src/engine/Service.cpp

```diff
@@ -241,65 +242,84 @@ void Service::writeJsonResult(const std::vector<std::string>& vars,
 }
 
 // ____________________________________________________________________________
-Result::Generator Service::computeResultLazily(
-    const std::vector<std::string> vars,
-    ad_utility::LazyJsonParser::Generator body, bool singleIdTable) {
-  LocalVocab localVocab{};
-  IdTable idTable{getResultWidth(), getExecutionContext()->getAllocator()};
+Result::LazyResult Service::computeResultLazily(
+    std::vector<std::string> vars, ad_utility::LazyJsonParser::Generator body,
+    bool singleIdTable) {
+  auto inputRange = ad_utility::CachingTransformInputRange(
+      std::move(body), [](auto& input) { return std::move(input); });
+  auto get = [service = this, vars = std::move(vars),
+              singleIdTable = singleIdTable, inputRange = std::move(inputRange),
+              localVocab = LocalVocab{},
+              idTable = IdTable{getResultWidth(),
+                                getExecutionContext()->getAllocator()},
+              rowIdx = size_t{0}, varsChecked = false, resultExists = false,
+              singleIdTableReturned =
+                  false]() mutable -> std::optional<Result::IdTableVocabPair> {
+    if (singleIdTableReturned) {
+      return std::nullopt;
+    }
+    try {
+      auto partJsonOpt = inputRange.get();
+      while (partJsonOpt.has_value()) {
+        const nlohmann::json& partJson = partJsonOpt.value();
+        if (partJson.contains("head")) {
+          AD_CORRECTNESS_CHECK(!varsChecked);
+          service->verifyVariables(partJson["head"],
+                                   inputRange.view().base().details());
```

Comment:
bind this to a variable right at the beginning of the lambda, i.e.
` const auto& details = .....base().details()` ,
such that the places where you use it become much simpler.